### ISSUE TITLE
`use_self` false positive on outer items

### ISSUE THREAD

**Issue by awused** [rust-lang/rust-clippy#13092]
### Summary

`use_self` doesn't work with outer items and the suggestion fails to compile.

### Lint Name

use_self

### Reproducer

I tried this code:

```rust
#![warn(clippy::use_self)]

#[derive(Default)]
struct MyStruct {}

impl MyStruct {
    thread_local! {
        static SPECIAL: RefCell<MyStruct> = RefCell::default();
    }
}
```

This warns on use_self, but attempting to change the signature to RefCell<Self> causes:

```
can't use `Self` from outer item
refer to the type directly here instead (rustc E0401)
```

### Version

```text
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: x86_64-unknown-linux-gnu
release: 1.79.0
LLVM version: 18.1.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
